File: Library/PaintedSkin/src/main/java/org/alee/component/skin/executor/BaseSkinExecutor.java
package org.alee.component.skin.executor;

import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.os.Looper;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;

import org.alee.component.skin.exception.ApplySkinException;
import org.alee.component.skin.pack.ResourcesType;
import org.alee.component.skin.service.Config;
import org.alee.component.skin.service.ThemeSkinService;
import org.alee.component.skin.util.ThreadUtils;

/**********************************************************
 *
 * @author: MY.Liu
 * @date: 2021/2/13
 * @description: xxxx
 *
 *********************************************************/
public abstract class BaseSkinExecutor<T extends View> implements ISkinExecutor {
    /**
     * {@link SkinElement}
     */
    private final SkinElement mElement;

    public BaseSkinExecutor(@NonNull SkinElement element) {
        mElement = element;
    }

    /**
     * 执行皮肤替换工作
     *
     * @param view 需要替换皮肤的View
     *
     * @throws ApplySkinException 换肤时出现的异常
     */
    @Override
    public final void execute(@NonNull View view) throws ApplySkinException {
        if (null == view) {
            return;
        }
        try {
            if (Thread.currentThread() == Looper.getMainLooper().getThread()) {
                applyThemeSkin((T) view, mElement);
            } else {
                Runnable runnable = new MainRunnable<>((T) view, mElement, this);
                // 此处需要注意 Handler.post 与View.post的区别
                // 使用View.post 由于非当前页面的View还没有AttachedToWindow,所以runnable 会被delay到 View AttachedToWindow
                if (Config.PerformanceMode.EXPERIENCE_FIRST == Config.getInstance().getPerformanceMode()) {
                    ThreadUtils.postOnMainThread(runnable);
                } else {
                    view.post(runnable);
                }
            }
        } catch (Throwable e) {
            throw new ApplySkinException(view, mElement, e);
        }
    }

    @Override
    public final String getSkinAttrName() {
        return mElement.getAttrName();
    }

    private void applyThemeSkin(@NonNull T view, @NonNull SkinElement element) {
        switch (element.getResourcesType()) {
            case ResourcesType.COLOR:
                ColorStateList colorStateList = ThemeSkinService.getInstance().getCurrentThemeSkinPack(view.getContext()).getColorStateList(element.getResourcesId());
                if (null != colorStateList) {
                    applyColor(view, colorStateList, element.getAttrName());
                    break;
                }
                applyColor(view, ThemeSkinService.getInstance().getCurrentThemeSkinPack(view.getContext()).getColor(element.getResourcesId()), element.getAttrName());
                break;
            case ResourcesType.DRAWABLE:
                applyDrawable(view, ThemeSkinService.getInstance().getCurrentThemeSkinPack(view.getContext()).getDrawable(element.getResourcesId()), element.getAttrName());
                break;
            case ResourcesType.MIPMAP:
                applyDrawable(view, ThemeSkinService.getInstance().getCurrentThemeSkinPack(view.getContext()).getMipmap(element.getResourcesId()), element.getAttrName());
                break;
            default:
                break;
        }
    }

    /**
     * 适用 颜色
     *
     * @param view           view
     * @param colorStateList 颜色状态集合
     * @param attrName       属性名称 例 textColor
     */
    protected abstract void applyColor(@NonNull T view, @NonNull ColorStateList colorStateList, @NonNull String attrName);

    /**
     * 适用 颜色
     *
     * @param view     view
     * @param color    色值
     * @param attrName 属性名称 例 textColor
     */
    protected abstract void applyColor(@NonNull T view, @ColorInt int color, @NonNull String attrName);

    /**
     * 适用 图片
     *
     * @param view     view
     * @param drawable 图片
     * @param attrName 属性名称 例 background
     */
    protected abstract void applyDrawable(@NonNull T view, @NonNull Drawable drawable, @NonNull String attrName);

    private static final class MainRunnable<T extends View> implements Runnable {
        private final T mView;
        private final SkinElement mElement;
        private final BaseSkinExecutor<T> mSkinExecutor;

        public MainRunnable(@NonNull T view, @NonNull SkinElement element, @NonNull BaseSkinExecutor<T> skinExecutor) {
            mView = view;
            mElement = element;
            mSkinExecutor = skinExecutor;
        }

        @Override
        public void run() {
            mSkinExecutor.applyThemeSkin(mView, mElement);
        }
    }

}

File: Library/PaintedSkin-Kotlin/src/main/java/org/alee/component/skin/core/executor/BaseSkinExecutor.kt
package org.alee.component.skin.core.executor

import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.MainThread
import org.alee.component.skin.ThemeSkinService
import org.alee.component.skin.constant.ResourcesType
import org.alee.component.skin.exception.ApplySkinException
import org.alee.component.skin.model.SkinAttribute
import org.alee.component.skin.template.ISkinExecutor
import org.alee.component.skin.util.ext.logI
import org.alee.component.skin.util.isOnMainThread

/**
 * 换肤执行器 - 基类
 *
 * <p> 详细描述
 *
 * @author MingYu.Liu
 * created in 2023/9/4
 *
 */
sealed class BaseSkinExecutor<T : View> : ISkinExecutor {
    /**
     * 执行换肤工作
     * @param view View 需要换肤的View
     * @param attribute SkinAttribute 换肤属性
     * @throws ApplySkinException 换肤时出现的异常
     */
    final override fun execute(view: View, attribute: SkinAttribute) {
        assertMainThread(view, attribute)
        try {
            applyThemeSkin(view as T, attribute)
        } catch (exception: Throwable) {
            throw ApplySkinException(view, attribute, exception)
        }
    }

    @Throws(Throwable::class)
    @MainThread
    private fun applyThemeSkin(view: T, attribute: SkinAttribute) {
        when (attribute.resourceType) {
            ResourcesType.COLOR -> {
                val colorStateList = ThemeSkinService.currentSkinPack.getColorStateList(attribute.resourceId)
                if (null == colorStateList) {
                    ThemeSkinService.currentSkinPack.getColor(attribute.resourceId)?.run {
                        applyColor(view, this, attribute.attrName)
                    }
                } else {
                    applyColor(view, colorStateList, attribute.attrName)
                }
            }

            ResourcesType.DRAWABLE -> ThemeSkinService.currentSkinPack.getDrawable(attribute.resourceId)?.run {
                applyDrawable(view, this, attribute.attrName)
            }

            ResourcesType.MIPMAP -> ThemeSkinService.currentSkinPack.getMipmap(attribute.resourceId)?.run {
                applyDrawable(view, this, attribute.attrName)
            }

            else -> {
                "Unsupported resource type: ${attribute.resourceType}".logI()
            }
        }
    }

    /**
     * 换肤 - ColorStateList
     * @param view T extends View 需要换肤的View
     * @param colorStateList ColorStateList 颜色状态集合
     * @param attrName String 属性名称 例 textColor
     * @throws Throwable 换肤时出现的异常
     */
    @Throws(Throwable::class)
    @MainThread
    protected abstract fun applyColor(view: T, colorStateList: ColorStateList, attrName: String)

    /**
     * 换肤 - Color
     * @param view T extends View 需要换肤的View
     * @param color Int 色值
     * @param attrName String 属性名称 例 textColor
     * @throws Throwable 换肤时出现的异常
     */
    @Throws(Throwable::class)
    @MainThread
    protected abstract fun applyColor(view: T, @ColorInt color: Int, attrName: String)

    /**
     * 换肤 - Drawable
     * @param view T extends View 需要换肤的View
     * @param drawable Drawable 图片
     * @param attrName String 属性名称 例 background
     * @throws Throwable 换肤时出现的异常
     */
    @Throws(Throwable::class)
    @MainThread
    protected abstract fun applyDrawable(view: T, drawable: Drawable, attrName: String)

    private fun assertMainThread(view: View, attribute: SkinAttribute) {
        if (isOnMainThread) {
            return
        }
        throw ApplySkinException(view, attribute, RuntimeException("Skin change must be performed on the main thread!"))
    }
}
